contracts/utils/BaalVotes.sol

```diff
@@ -0,0 +1,235 @@
+ *
+ * TODO alighn more with OZ ERC20Votes
+ */
+abstract contract BaalVotes is ERC20Permit {
+    using ECDSA for bytes32;
+
+    struct Checkpoint {
+        /*Baal checkpoint for marking number of delegated votes*/
+        uint32 fromTimeStamp; /*unix time for referencing voting balance*/
+        uint256 votes; /*votes at given unix time*/
+    }
+
+    // DELEGATE TRACKING
+    mapping(address => mapping(uint256 => Checkpoint)) public checkpoints; /*maps record of vote `checkpoints` for each account by index*/
+    mapping(address => uint256) public numCheckpoints; /*maps number of `checkpoints` for each account*/
+    mapping(address => address) public delegates; /*maps record of each account's `shares` delegate*/
+
+    // SIGNATURE HELPERS
+    mapping(address => uint256) public _nonces; /*maps record of states for signing & validating signatures*/
+    bytes32 constant DELEGATION_TYPEHASH =
+        keccak256("Delegation(address delegatee,uint256 nonce,uint256 expiry)");
+    bytes32 constant DOMAIN_TYPEHASH =
+        keccak256(
+            "EIP712Domain(string name,uint256 chainId,address verifyingContract)"
+        );
+
+    event DelegateChanged(
+        address indexed delegator,
+        address indexed fromDelegate,
+        address indexed toDelegate
+    ); /*emits when an account changes its voting delegate*/
+    event DelegateVotesChanged(
+        address indexed delegate,
+        uint256 previousBalance,
+        uint256 newBalance
+    ); /*emits when a delegate account's voting balance changes*/
+
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256 amount
+    ) internal virtual override {
+        super._beforeTokenTransfer(from, to, amount);
+
+        /*If recipient is receiving their first shares, auto-self delegate*/
+        if (balanceOf(to) == 0 && numCheckpoints[to] == 0 && amount > 0) {
+            delegates[to] = to;
+        }
+
+        _moveDelegates(delegates[from], delegates[to], amount);
+    }
+
+    /// @notice Delegate votes from user to `delegatee`.
+    /// @param delegatee The address to delegate votes to.
+    function delegate(address delegatee) public virtual {
+        _delegate(msg.sender, delegatee);
+    }
+
+    /// @notice Delegates votes from `signatory` to `delegatee` with EIP-712 signature.
+    /// @param delegatee The address to delegate 'votes' to.
+    /// @param nonce The contract state required to match the signature.
+    /// @param deadline The time at which to expire the signature.
+    /// @param signature The concatenated signature
+    //TODO
+    function delegateBySig(
+        address delegatee,
+        uint256 nonce,
+        uint256 deadline,
+        bytes calldata signature
+    ) public virtual {
+        bytes32 domainSeparator = keccak256(
+            abi.encode(
+                DOMAIN_TYPEHASH,
+                keccak256(bytes(name())),
+                // keccak256(bytes(name)),
```

Comment:
do we want to pass a version here?

if so, you should also add the version to the DOMAIN_TYPEHASH and the signPermit method (ussed in tests -- its in the src/ folder) - - otherwise, we can leave it here and i can update it on my branch - 

https://docs.metamask.io/guide/signing-data.html#sign-typed-data-v4

Reply:
yeah, I think it is worth doing that. lets get it in another pr because it may require test updates